/boot not mounted after reboot

Possibly because of https://github.com/aanderse/teraflops/blob/9b7cc368ffb7b3b3694286316db5dbe8b34d0b5f/nix/hcloud/default.nix#L42-L44 not specifying the `/boot` filesystem.

This causes subsequent deployments to be "wiped" after a reboot, as nixos cannot switch the profile in the bootloader.

In my case, boot is `/dev/sda15`.

Using hetzner cloud

```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";

    teraflops.url = "github:aanderse/teraflops";
    teraflops.inputs.nixpkgs.follows = "nixpkgs";
  };

  outputs = { nixpkgs, teraflops, ... }:
    let
      system = "x86_64-linux";

      hcloud-servers = [ "aries" "centaurus" ];
      hcloud-config = {
        server_type = "ccx13";
        location = "nbg1";
      };

      server-config = { pkgs, lib, ... }: {
        environment.systemPackages = with pkgs; [ htop nano zulu17 ];

        networking.firewall.enable = false;

        services.pufferpanel = {
          enable = true;
          environment = { PUFFER_PANEL_ENABLE = "false"; };
          extraPackages = with pkgs; [ bash curl gawk gnutar gzip ];
          package = pkgs.buildFHSEnv {
            name = "pufferpanel-fhs";
            runScript = lib.getExe pkgs.pufferpanel;
            targetPkgs = pkgs': with pkgs'; [ icu openssl zlib ];
          };
        };
      };
    in {
      teraflops = (

        ({
          imports = [ teraflops.modules.hcloud ];
          meta = { nixpkgs = import nixpkgs { system = system; }; };
        })

        // (builtins.listToAttrs (builtins.map (s: {
          name = s;
          value = { pkgs, lib, ... }:
            ({
              deployment.targetEnv = "hcloud";
              deployment.hcloud = hcloud-config;
            } // (server-config { inherit pkgs lib; }));
        }) hcloud-servers))

        # TODO: Other providers/manual servers
teraflops nix apply --reboot
      );
    };
}
```

The reason I suspect it's the boot not mounting, is the following error when changing the config, and running `teraflops nix apply --reboot`:

```
 teraflops nix apply --reboot
warning: Git tree '/home/polarizedions/programming/cosmos' is dirty
[INFO ] Using configuration: /tmp/teraflops.6q_uof5o/hive.nix
warning: Git tree '/home/polarizedions/programming/cosmos' is dirty
[INFO ] Enumerating nodes...
warning: Git tree '/home/polarizedions/programming/cosmos' is dirty
warning: Git tree '/home/polarizedions/programming/cosmos' is dirty
[INFO ] Selected all 1 nodes.
      🕚 33s 1 running, 5 succeeded
aries ✅ 4s Evaluated aries
      🕙 42s 1 running, 5 succeeded
aries ✅ 4s Evaluated aries
      🕑 47s 1 running, 5 succeeded
aries ✅ 4s Evaluated aries
      ❌ 53s Failed: Unexpected active profile: Profile(StorePath("/nix/store/2idgdn6h5gcwg0bk9ahpf6qf7fcmwzxk-nixos-system-aries-24.05pre-git"))
aries ✅ 4s Evaluated aries
aries ✅ 1s Built "/nix/store/a60y9mv85hc6dgvr0ranrww2pfrj09v8-nixos-system-aries-24.05pre-git"
aries ✅ 9s Pushed system closure
aries ✅ 5s Will be activated next boot
aries ❌ 34s Reboot failed: Unexpected active profile: Profile(StorePath("/nix/store/2idgdn6h5gcwg0bk9ahpf6qf7fcmwzxk-nixos-system-aries-24.05pre-git"))                                             
[ERROR] Failed to complete requested operation - Last 1 lines of logs:
[ERROR]  failure) Unexpected active profile: Profile(StorePath("/nix/store/2idgdn6h5gcwg0bk9ahpf6qf7fcmwzxk-nixos-system-aries-24.05pre-git"))
[ERROR] Failed to reboot aries - Last 4 lines of logs:
[ERROR]  created)
[ERROR]    state) Running
[ERROR]  message) Waiting for reboot
[ERROR]  failure) Unexpected active profile: Profile(StorePath("/nix/store/2idgdn6h5gcwg0bk9ahpf6qf7fcmwzxk-nixos-system-aries-24.05pre-git"))
[ERROR] -----
[ERROR] Operation failed with error: Unexpected active profile: Profile(StorePath("/nix/store/2idgdn6h5gcwg0bk9ahpf6qf7fcmwzxk-nixos-system-aries-24.05pre-git"))
```

--- a/examples/hcloud/flake.nix
+++ b/examples/hcloud/flake.nix
@@ -22,7 +22,7 @@
           packages = [
             colmena
             jq
-            (terraform.withPlugins (p: [ p.hcloud p.tls ]))
+            (terraform.withPlugins (p: [ p.hcloud p.ssh p.tls ]))
             teraflops.packages.${system}.default
           ];
         };

--- a/nix/hcloud/default.nix
+++ b/nix/hcloud/default.nix
@@ -3,7 +3,11 @@ let
   nodes' = lib.filterAttrs (_: node: node.targetEnv == "hcloud") (outputs.teraflops.nodes or {});
 in
 {
-  defaults = { modulesPath, name, config, pkgs, lib, ... }: with lib; {
+  defaults = { name, config, pkgs, lib, ... }: with lib;
+  let
+    remoteCfg = builtins.fromJSON resources.ssh_resource.${name}.result;
+  in
+  {
     options.deployment.hcloud = mkOption {
       type = with types; nullOr (submodule {
         freeformType = (pkgs.formats.json {}).type;
@@ -36,13 +40,20 @@ in
         then resources.hcloud_server.${name}.ipv4_address
         else tf.ref "hcloud_server.${name}.ipv4_address";
 
-      boot.loader.grub.device = "/dev/sda";
+      boot.loader.grub = remoteCfg.boot.loader.grub or {
+        device = "/dev/sda";
+      };
       boot.initrd.kernelModules = [ "nvme" ];
 
-      fileSystems."/" = {
-        fsType = "ext4";
-        device = "/dev/sda1";
-      };
+      fileSystems = mkMerge [
+        (remoteCfg.fileSystems or {})
+        {
+          "/" = {
+            fsType = "ext4";
+            device = "/dev/sda1";
+          };
+        }
+      ];
 
       networking.defaultGateway = "172.31.1.1";
       
@@ -93,6 +104,10 @@ in
         source = "hetznercloud/hcloud";
         version = ">= 1.44.0";
       };
+      ssh = {
+        source = "loafoe/ssh";
+        version = ">= 2.7.0";
+      };
     };
   };
 
@@ -102,6 +117,41 @@ in
     in
     {
       hcloud_server = mapAttrs (_: node: node.config.deployment.hcloud) nodes';
+      ssh_resource = mapAttrs (name: node: {
+        user = node.config.deployment.targetUser;
+        host = node.config.deployment.targetHost;
+        port = mkIf (node.config.deployment.targetPort != null) node.config.deployment.targetPort;
+        private_key = mkIf node.config.deployment.provisionSSHKey (tf.ref "tls_private_key.teraflops.private_key_openssh");
+
+        # https://github.com/elitak/nixos-infect/blob/5ef3f953d32ab92405b280615718e0b80da2ebe6/nixos-infect#L219
+        file = {
+          destination = "/tmp/teraflops.sh";
+          permissions = "0700";
+          content = ''
+            #!/usr/bin/env bash
+
+            esp=""
+            for d in /boot/EFI /boot/efi /boot; do
+              [[ ! -d "$d" ]] && continue
+              [[ "$d" == "$(df "$d" --output=target | sed 1d)" ]] \
+                && esp="$(df "$d" --output=source | sed 1d)" \
+                && break
+            done
+            [[ -z "$esp" ]] && { echo '{}'; exit 0; }
+            for uuid in /dev/disk/by-uuid/*; do
+              [[ $(readlink -f "$uuid") == "$esp" ]] && echo "{\"boot\": {\"loader\": {\"grub\": {\"efiSupport\": true, \"efiInstallAsRemovable\": true, \"device\": \"nodev\"}}}, \"fileSystems\": {\"/boot\": {\"device\": \"$uuid\", \"fsType\": \"vfat\"}}}" && exit 0
+            done
+          '';
+        };
+
+        commands = [
+          "/tmp/teraflops.sh"
+        ];
+
+        depends_on = [
+          "hcloud_server.${name}"
+        ];
+      }) nodes';
     };
 } // lib.mapAttrs (_: node: { modulesPath, ... }: {
   imports = [ "${modulesPath}/profiles/qemu-guest.nix" ];

